edgelet/edgelet-docker/src/migc_persistence.rs

```diff
@@ -0,0 +1,312 @@
+
+    Ok(image_map)
+}
+
+fn write_images_with_timestamp(
+    state_to_persist: &HashMap<String, Duration>,
+    temp_file: String,
+    filename: String,
+) -> Result<(), Error> {
+    // write to a temp file and then rename/overwrite to image persistence file
+    let mut file =
+        std::fs::File::create(temp_file.clone()).expect("Could not create images under /tmp");
+
+    for (key, value) in state_to_persist {
+        let image_details = format!("{} {}\n", key, value.as_secs());
+        let res = write!(file, "{}", image_details);
+        if res.is_err() {
+            let msg = format!(
+                "Could not write image:{} with timestamp:{} to store",
+                key,
+                value.as_secs()
+            );
+            log::error!("{}", msg);
+            return Err(Error::FileOperation(msg));
+        }
+    }
+
+    // add retries?
+    match fs::rename(temp_file, filename) {
+        Ok(_) => {},
+        Err(_) => return Err(Error::FileOperation(
+            "Could not update auto-prune data; next run may try to delete images that are no longer present on device".to_string(),
+        )),
+    };
+
+    Ok(())
+}
+
+fn process_state(
+    mut image_map: HashMap<String, Duration>,
+    running_modules: Vec<(
+        DockerModule<http_common::Connector>,
+        edgelet_core::ModuleRuntimeState,
+    )>,
+    min_age: Duration,
+) -> (HashMap<String, Duration>, HashMap<String, Duration>) {
+    let current_time = std::time::SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Could not get EPOCH time");
+
+    let mut carry_over: HashMap<String, Duration> = HashMap::new(); // all images to NOT be deleted by MIGC in this run
+
+    // then, based on ID, keep track of images currently being used (in map: carry_over)
+    for module in running_modules {
+        let key = module
+            .0
+            .config()
+            .image_hash()
+            .expect("Could not get image id from module");
+
+        // Since the images are currently being used, we update the timestamp to the current time
+        // This avoids the case where a container crash just as MIGC is kicking off removes a needed image
+
+        // TODO: Do we need to trim the key here?
+        carry_over.insert(key.to_string(), current_time);
+    }
+
+    // track entries younger than min age
+    for (key, value) in &image_map {
+        if current_time.as_secs() - value.as_secs() < min_age.as_secs() {
+            carry_over.insert(key.to_string(), *value);
+        }
+    }
+
+    // clean up image map to make sure entries that need to be preserved are not removed
+    for key in carry_over.keys() {
+        image_map.remove(key);
+    }
+
+    (image_map, carry_over)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{collections::HashMap, time::Duration};
+
+    use crate::migc_persistence::get_images_with_timestamp;
+
+    use super::write_images_with_timestamp;
+
+    const TEMP_FILE: &str = "/tmp/images";
```

Comment:
Writing to tmp will be a problem for tests:
```
thread 'thread 'migc_persistence::tests::test_get_write_images_with_timestamp' panicked at 'test migc_persistence::tests::test_file_rename_write_images_with_timestamp ... ok
assertion failed: result.is_ok()migc_persistence::tests::test_panic_write_images_with_timestamp', ' panicked at 'edgelet-docker/src/migc_persistence.rs:Could not create images under /tmp: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', edgelet-docker/src/migc_persistence.rs:171300::509
```

How did you get this to work for you?